agent/consul/multilimiter/multilimiter.go

```diff
@@ -0,0 +1,277 @@
+	limiters := atomic.Pointer[radix.Tree]{}
+	configs := atomic.Pointer[radix.Tree]{}
+	config := atomic.Pointer[Config]{}
+	config.Store(&c)
+	limiters.Store(radix.New())
+	configs.Store(radix.New())
+	if c.ReconcileCheckLimit == 0 {
+		c.ReconcileCheckLimit = 30 * time.Millisecond
+	}
+	if c.ReconcileCheckInterval == 0 {
+		c.ReconcileCheckLimit = 1 * time.Second
+	}
+	chLimiter := make(chan *limiterWithKey, 100)
+	m := &MultiLimiter{limiters: &limiters, defaultConfig: &config, limitersConfigs: &configs, limiterCh: chLimiter}
+
+	return m
+}
+
+// Run the cleanup routine to remove old entries of Limiters based on ReconcileCheckLimit and ReconcileCheckInterval.
+func (m *MultiLimiter) Run(ctx context.Context) {
+	m.once.Do(func() {
+		go func() {
+			writeTimeout := 10 * time.Millisecond
+			limiters := m.limiters.Load()
+			txn := limiters.Txn()
+			waiter := time.NewTicker(writeTimeout)
+			wt := tickerWrapper{ticker: waiter}
+			defer waiter.Stop()
+			for {
+				if txn = m.runStoreOnce(ctx, wt, txn); txn == nil {
+					return
+				}
+			}
+		}()
+		go func() {
+			waiter := time.NewTimer(0)
+			for {
+				c := m.defaultConfig.Load()
+				waiter.Reset(c.ReconcileCheckInterval)
+				select {
+				case <-ctx.Done():
+					waiter.Stop()
+					return
+				case now := <-waiter.C:
+					m.reconcileLimitedOnce(now, c.ReconcileCheckLimit)
+				}
+			}
+		}()
+	})
+
+}
+
+// todo: split without converting to a string
+func splitKey(key []byte) ([]byte, []byte) {
+
+	ret := bytes.SplitN(key, []byte(separator), 2)
+	if len(ret) != 2 {
+		return []byte(""), []byte("")
+	}
+	return ret[0], ret[1]
+}
+
+// Allow should be called by a request processor to check if the current request is Limited
+// The request processor should provide a LimitedEntity that implement the right Key()
+func (m *MultiLimiter) Allow(e LimitedEntity) bool {
+	prefix, _ := splitKey(e.Key())
+	limiters := m.limiters.Load()
+	l, ok := limiters.Get(e.Key())
+	now := time.Now()
+	unixNow := time.Now().UnixMilli()
+	if ok {
+		limiter := l.(*Limiter)
+		if limiter.limiter != nil {
+			limiter.lastAccess.Store(unixNow)
+			return limiter.limiter.Allow()
+		}
+	}
+
+	configs := m.limitersConfigs.Load()
+	c, okP := configs.Get(prefix)
+	var config = &m.defaultConfig.Load().LimiterConfig
+	if okP {
+		prefixConfig := c.(*LimiterConfig)
+		if prefixConfig != nil {
+			config = prefixConfig
+		}
+	}
+	limiter := &Limiter{limiter: rate.NewLimiter(config.Rate, config.Burst)}
+	limiter.lastAccess.Store(unixNow)
+	m.limiterCh <- &limiterWithKey{l: limiter, k: e.Key(), t: now}
+	return limiter.limiter.Allow()
+}
+
+type ticker interface {
+	Ticker() <-chan time.Time
+}
+
+type tickerWrapper struct {
+	ticker *time.Ticker
+}
+
+func (t tickerWrapper) Ticker() <-chan time.Time {
+	return t.ticker.C
+}
+
+func (m *MultiLimiter) runStoreOnce(ctx context.Context, waiter ticker, txn *radix.Txn) *radix.Txn {
+	select {
+	case <-waiter.Ticker():
+		tree := txn.Commit()
+		m.limiters.Store(tree)
```

Comment:
I think we need to make this a `CompareAndSwap` operation (like we do in `reconcileLimitedOnce`).

My thinking is that if the reconciliation loop takes a long time, there's a pretty good likelihood of the store/commit loop undoing its changes.

```
R = Reconciliation loop
S = Store loop
# = Transaction in-flight

      0ms   10ms   20ms   30ms   40ms   50ms
+---+----------------------------------------+
| R |        |###################|           |
+---+----------------------------------------+
| S |                      |########|        |
+---+----------------------------------------+
```

In the pathological case, the right pattern of load could prevent us from _ever_ re-configuring or cleaning up limiters.

What do you think @dhiaayachi?

Reply:
Actually thinking about it a little more, `reconcileLimitedOnce` taking a long time means it'll probably end up failing at the `CompareAndSwap` frequently anyway because the tree would've been overwritten 🤔 

I wonder if we should accept the performance overhead of just not using transactions. We don't need them for atomicity, for example.

Reply:
We discussed this with @boxofrad and we agreed that we don't need to fully solve this but try to reduce the surface/impact of this happening. As config updates and clean-up would happen periodically and would end up converging under the right circonstances. I will start another PR with a way to do that after this is merged.